[FEAT] 마이페이지 > 주소지 등록 기능 추가 (#45)

mochafreddo opened the pull request:
#### 개요

- 마이페이지 > 설정 > 주소지 관리 기능을 위한 백엔드 API 및 도메인 구조를 추가합니다.
- 사용자는 여러 주소지를 가질 수 있으며, 한 개 이상의 주소지가 있을 경우 반드시 하나의 기본 주소지를 갖도록 설계하였습니다.
- 주소지 정보는 이름, 연락처, 주소명, 기본 주소지 여부, 주소1(기본), 주소2(상세)로 구성됩니다.
- 확장성과 향후 기능 추가를 고려하여 테이블 및 엔티티를 설계하였습니다.

---

#### 주요 변경사항

1. **Address 도메인 신규 추가**
    
   - **`AddressController.java`**: 주소지 생성 (`POST /api/v1/addresses`) 및 목록 조회 (`GET /api/v1/addresses`) API를 제공합니다.
   - **`AddressSaveRequest.java`**, **`AddressResponse.java`**: 주소지 저장 요청/응답 DTO입니다.
   - **`UserAddress.java`**: 주소지 엔티티로 JPA Auditing이 적용되어 `createdAt` 및 `updatedAt` 필드가 포함됩니다. 또한, `user_address` 테이블에 `user_id` 컬럼을 이용한 인덱스(`idx_user_id`)가 추가되었습니다1.
   - **`UserAddressRepository.java`**: 유저별 주소지 목록 조회 (`findByUser`)를 지원합니다2.
   - **`AddressService.java`**: 주소지 저장 및 조회 관련 비즈니스 로직을 처리합니다.

2. **DroniUserRepository 확장**
    
   - `findByOauthId` 메서드를 추가하여 OAuth2 principal 기반으로 유저를 조회하는 기능을 지원합니다.

3. **DroniUser 알림 설정 필드 변경**
    
   - 기존의 `notificationEnabled`와 `marketingEnabled` 필드가 더 세분화된 알림 설정 필드로 변경되었습니다.
   - **`appPushNotificationEnabled`**
   - **`kakaoTalkNotificationEnabled`**
   - **`chatNotificationEnabled`**
   - **`estimateBidNotificationEnabled`**

4. **테스트 코드 추가**

   - **`AddressControllerTest`**, **`AddressServiceTest`**: API 및 서비스 레이어 단위 테스트가 추가되었습니다.
   - 기본 주소지 정책, 저장/조회 케이스를 검증하는 테스트가 포함됩니다.
   - 인증 정보가 없는 경우 404 Not Found 예외를 반환하는 테스트가 추가되었습니다.

5. **보안 테스트 의존성 추가**

   - `build.gradle` 파일에 **`org.springframework.security:spring-security-test`** 의존성이 추가되었습니다.

---

#### 상세 구현 및 정책

- **기본 주소지 정책**
    
    - 첫 주소지는 무조건 기본 주소지로 저장됩니다.
    - 새 주소지가 기본으로 지정되면 기존 기본 주소지는 해제됩니다.
    - 기본 주소지가 없는 상태에서 새 주소지가 기본이 아니면 자동으로 기본 주소지로 승격됩니다.

- **확장성**
    
    - `UserAddress` 엔티티에 JPA Auditing 필드(`createdAt`, `updatedAt`) 포함되어 있습니다.
    - 추후 주소지 삭제, 수정, 기타 정보 추가가 용이하도록 설계되었습니다.

---

#### API 명세

- **POST `/api/v1/addresses`**
    
    - 주소지 저장 (인증 필요).
    - Request: `AddressSaveRequest`
    - Response: `AddressResponse`

- **GET `/api/v1/addresses`**
    
    - 유저의 주소지 목록 조회 (인증 필요).
    - Response: `List<AddressResponse>`

---

#### 테스트

- 서비스 및 컨트롤러 단위테스트 포함.
- 기본 주소지 정책, 저장/조회 정상동작 검증.
- 존재하지 않는 사용자로 요청 시 예외 처리 검증.

---

#### 관련 이슈

- #45

---

#### 변경 파일

- `build.gradle`
- `AddressController.java`
- `AddressResponse.java`    
- `AddressSaveRequest.java`
- `UserAddress.java`
- `UserAddressRepository.java`
- `AddressService.java`
- `DroniUser.java`
- `DroniUserQuerydslRepository.java`
- `DroniUserQuerydslRepositoryImpl.java`
- `OAuth2UserPrincipal.java`
- `AddressControllerTest.java`
- `AddressServiceTest.java`
- `DroniUserRepositoryTest.java`
- `test-expert.sql`
- `test-user-set.sql`

---

리뷰 부탁드립니다. 🙇‍♂️

mochafreddo committed:
feat(address): 사용자 주소 도메인 엔티티, DTO, 서비스, 리포지토리 추가

- UserAddress 엔티티 및 AddressSaveRequest, AddressResponse DTO 추가
- UserAddressRepository JPA 리포지토리 구현
- AddressService에서 주소 저장 및 조회 로직 구현
- 첫 주소는 무조건 기본 주소로 지정, 기본 주소 변경 시 기존 기본 주소 해제 처리

Refs: #45

mochafreddo committed:
feat(address): 주소지 API 컨트롤러(AddressController) 추가

- 주소지 생성 및 조회를 위한 AddressController 구현
- 인증된 사용자의 OAuth2UserPrincipal을 활용하여 주소지 저장 및 목록 조회 기능 제공
- Swagger 문서화 및 SecurityRequirement 적용

Refs: #45

mochafreddo committed:
feat: DroniUserQuerydslRepository에 findByOauthId 메서드 추가

- DroniUserQuerydslRepository 및 구현체에 findByOauthId(String oauthId) 메서드 추가
- 기존 findByUserOauth2Id와 유사하게 oauthId로 사용자 조회 가능하도록 구현
- 코드 정렬 및 불필요한 공백 정리

Refs: #45

mochafreddo committed:
test: 주소지 API 단위 테스트(AddressControllerTest, AddressServiceTest) 추가

- AddressControllerTest: 주소 저장 및 목록 조회 요청에 대한 컨트롤러 단위 테스트 추가
- AddressServiceTest: 주소 저장 로직(기본 주소 지정/변경/유지) 및 목록 조회 서비스 단위 테스트 추가
- Mock 객체를 활용하여 서비스/컨트롤러의 주요 동작 시나리오 검증

Refs: #45

mochafreddo commented:
```sql
CREATE TABLE user_address (
    address_id SERIAL PRIMARY KEY,
    address_name VARCHAR(100) NOT NULL,
    is_primary BOOLEAN NOT NULL DEFAULT FALSE,
    recipient_name VARCHAR(50) NOT NULL,
    contact_number VARCHAR(30) NOT NULL,
    address1 VARCHAR(255) NOT NULL,
    address2 VARCHAR(255),
    user_id BIGINT NOT NULL,
    created_at TIMESTAMP WITHOUT TIME ZONE DEFAULT CURRENT_TIMESTAMP,
    updated_at TIMESTAMP WITHOUT TIME ZONE DEFAULT CURRENT_TIMESTAMP,
    CONSTRAINT fk_user_address_user
        FOREIGN KEY (user_id) REFERENCES droni_user(user_id)
);

CREATE INDEX idx_user_address_user_id ON user_address(user_id);
```

TaeTaeKim commented on src/main/java/droni/backend/api/address/dto/AddressResponse.java:
```diff
@@ -0,0 +1,44 @@
+    private final LocalDateTime updatedAt;
+
+    @Builder
+    public AddressResponse(Long addressId, String addressName, boolean isPrimary,
```
요부분은 모든 파라미터가 전부 들어가는거 같은데 Class 레벨에 Builder를 안붙이신 이유가 있을까요?

mochafreddo replied:
아니요. Gemini CLI가 작성했는데 일반적인 방식대로 클래스 레벨에 작성하는 편이 좋을 것 같네요. 추가 커밋 올릴게요 :)

TaeTaeKim replied:
이 AllargContructor를 그냥 삭제하는 건 어떨까요??

TaeTaeKim commented on src/main/java/droni/backend/api/address/controller/AddressController.java:
```diff
@@ -0,0 +1,48 @@
+    private final AddressService addressService;
+    private final DroniUserRepository droniUserRepository;
+
+    @Operation(summary = "주소지 생성", security = @SecurityRequirement(name = "Authorization"))
```
Swagger.config 에 글로벌하게 설정되있기도 하고 이렇게 개별로 설정하면 글로벌 설정을 덮어쓰게 되어 첨부된 사진처럼 보이고 Swagger 상에서 테스트가 불가능합니다 ㅠ

<img width="1498" height="218" alt="Image" src="https://github.com/user-attachments/assets/1ddb4cf8-1bb2-4ef9-9a22-7ed4cc71df89" />

mochafreddo replied:
아! 네! 수정 후 커밋했습니다!

mochafreddo committed:
feat(address): AddressResponse에 @Builder 어노테이션 추가

AddressResponse DTO 클래스에 @Builder 어노테이션을 클래스 레벨에 추가하여 빌더 패턴을 사용할 수 있도록 개선하였습니다. 기존 생성자에 있던 @Builder 어노테이션은 제거하였습니다.

Refs: #45, #47

mochafreddo committed:
refactor(address): Swagger 보안 어노테이션 개별 설정 제거

SwaggerConfig에서 글로벌 보안 설정을 적용하고 있으므로,
AddressController의 메서드별 @SecurityRequirement 어노테이션을 제거하여
Swagger UI에서 인증 및 테스트가 정상적으로 동작하도록 개선함.

Refs: #45, #47

TaeTaeKim commented on src/main/java/droni/backend/api/address/entity/UserAddress.java:
```diff
@@ -0,0 +1,71 @@
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
+
+    public UserAddress(Long addressId, String addressName, boolean isPrimary, String recipientName,
```
해당부분은 @AllArgConstructor 로 하면 조금 더 엔티티 코드가 정리될 듯 합니다!

mochafreddo replied:
이 부분에 대해서 Gemini로 검토요청을 했는데

---

`@AllArgsConstructor` 사용은 JPA 엔티티의 특성을 고려할 때 몇 가지 부작용을 낳을 수 있어 현재 코드를 유지하거나 다른 방식으로 개선하는 것이 더 좋다고 생각합니다.

#### `@AllArgsConstructor` 사용을 권장하지 않는 이유

1. **JPA 관리 필드 포함**: `@AllArgsConstructor`는 클래스의 모든 필드를 포함하는 생성자를 만듭니다. `UserAddress` 엔티티의 경우, 데이터베이스가 자동 생성하는 `addressId`와 `AuditingEntityListener`가 관리하는 `createdAt`, `updatedAt` 필드까지 생성자에 포함됩니다. 이 필드들은 개발자가 직접 값을 할당하는 것을 지향하지 않으므로, 생성자에 포함되는 것은 바람직하지 않습니다.
2. **예상치 못한 동작 가능성**: JPA 구현체(예: Hibernate)는 엔티티 객체를 프록시로 감싸서 관리할 때 특정 생성자를 사용하는데, `@AllArgsConstructor`가 의도치 않은 상호작용을 일으킬 수 있습니다.
3. **취약한 API**: 엔티티에 필드가 추가되거나 삭제될 때마다 생성자의 파라미터가 변경되어, 이 생성자를 사용하는 모든 코드에 영향을 주게 됩니다.

---

라고 하는데 어떻게 할까요? 🤔
저는 아직 이 부분에 대해서 판단하기 어려워서 태윤님 지침에 따르겠습니다 :)

TaeTaeKim replied:
해당 말처럼 사실 엔티티는 AllArgConstructor 를 쓰지 않습니다 마찬가지로 생성자 builder 는 private 로 하고 보통 엔티티생성은 정적팩토리메소드를 만들고 내부에서 builder 룰 사용하는게 제일 안전합니다

mochafreddo replied:
반영했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/a1034ad1de55b43ed9b0c4b05cd6c9dbbc43ad81)

TaeTaeKim commented on src/main/java/droni/backend/api/address/repository/UserAddressRepository.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.List;
+
+public interface UserAddressRepository extends JpaRepository<UserAddress, Long> {
+    List<UserAddress> findByUser(DroniUser user);
```
유저로 많이 조회하는 거면 user_id 에 인덱스를 주는 건 어떨까요?

mochafreddo replied:
반영했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/bc43786f85ef6ef73a52fd06856493e1fa629b38)

TaeTaeKim commented on src/main/java/droni/backend/api/address/service/AddressService.java:
```diff
@@ -0,0 +1,62 @@
+
+    public UserAddress saveAddress(Long userId, AddressSaveRequest request) {
+        DroniUser user = droniUserRepository.findById(userId)
+                .orElseThrow(() -> new RuntimeException("User not found"));
```
해당 부분은 저희 Droni 안에 `BaseException` 을 추가 정의하거나 미리 정의된 BaseException 자식 클래스가 있으니 참조하셔서 해당 Excetpion을 사용하시면 좋을 듯 싶습니다!

mochafreddo replied:
반영했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/837f6662c7fbdafaabb049c3f80575eb4860b691)

TaeTaeKim commented on src/test/java/droni/backend/api/address/controller/AddressControllerTest.java:
제 경험상으론, 해당 테스트들은 MockMVC를 이용한 테스트를 사용하시면 더욱 실제 요청같은 테스트를 하실 수 있습니다. 한번 찾아보시면 좋을 것 같아요! 물론 현재 테스트도 괜찮습니다 👍

mochafreddo replied:
기본적인 MockMvc 활용 슬라이스 테스트로 변경했습니다!

TaeTaeKim commented on src/main/java/droni/backend/api/address/service/AddressService.java:
```diff
@@ -0,0 +1,62 @@
+                    isNewAddressPrimary = true;
+                }
+            }
+        }
```
해당 부분은 if중첩이 들어가서 조금 복잡해져보이는데 이런 방향은 어떠세요??

로직을 보니까 "주소가 하나라도 있으면" -> "무조건 primary는 존재" 인 것 같습니다. 이는 "primary 가 없음" == "등록된 주소가 없음" 이기 떄문에 else문 안에있는 `if (!hasExistingPrimary)` 는 들어올 수 없는 구조일 것 같습니다.

고로 primary 주소가 있는지만 체크하면 될 것같습니다.

pseudo 코드를 적자면

request 가 primary 면 -> 기존 값들은 다 pirmary false로 바꾸면서 저장( 기존 pirmary 존재유무는 상관없음)
request 가 not primary이면 -> primary 유무 체크후 저장

이렇게 하면 if 문 내부에서도 stream을 도는 것도 방지할 수 있고 2중 if 분기는 막을 수 있을 것 같습니다.

mochafreddo replied:
반영했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/64e967e8300a9e0205d7a4afced1702809b8103a)

mochafreddo committed:
fix(address): 유저 조회 실패 시 DroniNotFoundException 사용으로 예외 처리 일관성 개선

AddressService에서 유저를 찾지 못할 때 기존 RuntimeException 대신 DroniNotFoundException을 사용하도록 변경하여, 글로벌 예외 처리 정책에 맞게 일관성을 높였습니다.

Refs: #45, #47

mochafreddo committed:
feat(address): 유저 조회 실패 시 DroniNotFoundException 사용

AddressController에서 유저를 찾지 못할 때 RuntimeException 대신 DroniNotFoundException을 사용하도록 변경하였습니다.
이로써 예외 처리 일관성과 글로벌 예외 핸들러 적용이 보장됩니다.

Refs: #45, #47

mochafreddo committed:
feat(address): user_id 컬럼에 인덱스 추가

UserAddress 엔티티의 user_id 컬럼에 인덱스를 추가하여 유저 기준 조회 성능을 개선함.
리뷰어 피드백을 반영한 변경사항임.

Refs: #45, #47

mochafreddo committed:
Merge branch 'develop' into feat/mypage-add-user-address-45

mochafreddo committed:
test: AddressService 예외 케이스 유닛 테스트 추가

- 사용자를 찾을 수 없는 경우 saveAddress, getUserAddresses에서 DroniNotFoundException이 발생하는지 검증하는 테스트를 추가함
- addressService의 예외 처리 신뢰성 강화

Refs: #45, #47

mochafreddo committed:
test: 주소 서비스 단위 테스트 리팩터링 및 시나리오 명확화

- AddressServiceTest의 테스트 메서드명을 도메인/시나리오 기반으로 개선
- ArgumentCaptor 활용 등 검증 방식 개선
- 사용자 미존재/주소 미존재 등 예외 및 엣지케이스 테스트 추가
- DisplayName을 요구사항 기반 코드(예: SAVE-SUCCESS-001 등)로 통일
- 불필요한 중복 코드 제거 및 가독성 향상

Refs: #45, #47

TaeTaeKim commented on src/main/java/droni/backend/api/address/controller/AddressController.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Operation(summary = "주소지 생성")
+    @PostMapping
+    public ResponseEntity<AddressResponse> saveAddress(
```
Controller 에 RestController 어노테이션이 있으면 반환타입을 `ResponseEntity` 로 감싸지 않고 바로 Return 해도 가능 합니다. 보통 Controller 에서 ResponseEntity의 여러 반환 Status등을 수정할 때 사용하지만 해당 Controller 에서는 없애면 조금 더 깔끔한 코드가 될 것 같습니다!

mochafreddo replied:
반영했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/ee526ec52d719f51032654fc4a350cd0df3ce6f0) (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/517364bc8fe1b160b502631b16cd852b7b9a1e9f)

TaeTaeKim commented on src/test/java/droni/backend/api/address/service/AddressServiceTest.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Test
+    @DisplayName("SAVE-SUCCESS-004: 기존에 기본 주소 없을 시, 새 주소는 기본 주소로 자동 설정")
```
아직 삭제 로직은 없지만 서비스 로직을 체크했을 때는 이 상황에 들어올 수 있나요??

로직 상 지향하는 방향이 무조건 primary 하나는 있는 것으로 보여서요.

mochafreddo replied:
맞습니다. `primary`는 하나인 것으로 임시기획/구현했는데, 현재는 이 테스트케이스가 해당이 없으나 추후 삭제 로직이 생기면 이에 대한 로직을 정비해야할 것 같아요. 현재는 방어코드 구현 차원에서 테스트케이스까지 포함되었다고 보시면 됩니다. 예외 케이스에 대한 상기 수준의 코드인데, 필요 없으면 제거할까요?

mochafreddo replied:
아래 리뷰와 함께 반영했습니다. (https://github.com/DronicaPlatformService/droni-backend/commit/70ce9e58e3b028eb479951422572bedbf0f08a64)

TaeTaeKim commented on src/test/java/droni/backend/api/address/service/AddressServiceTest.java:
```diff
@@ -0,0 +1,198 @@
+
+    @Test
+    @DisplayName("GET-SUCCESS-001: 주소 목록 조회 성공")
+    void getUserAddresses_shouldReturnAddressList() {
```
이 케이스 경우 만드셨으니 상관없지만, 보통 JPA의 조회 같은 경우는 가져와서 특별한 로직이 있지 않는 이상 JPA 라이브러리의 동작을 믿는 것으로 생략한다고도 합니다:)

TaeTaeKim commented on src/test/java/droni/backend/api/address/service/AddressServiceTest.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Test
+    @DisplayName("GET-SUCCESS-002: 주소가 없는 사용자의 목록 조회 시, 빈 리스트 반환")
```
해당 부분은 서비스의 빈 리스트 반환도 중요하지만 "Repository layer 가 빈 리스트를 잘 반환하냐" 가 중요하고 보통 JPA 를 사용하여 Collection Type 으로 반환시에는 항상 빈 Collection 으로 반환합니다.!

mochafreddo replied:
반영했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/70ce9e58e3b028eb479951422572bedbf0f08a64)

mochafreddo committed:
feat(address): AddressSaveRequest에 @Setter 어노테이션 추가

AddressSaveRequest DTO에 @Setter를 적용하여 테스트 및 서비스 계층에서 필드 값 설정이 가능하도록 개선하였습니다.

Refs: #45, #47

mochafreddo committed:
test: 주소지 컨트롤러 테스트를 WebMvcTest 기반으로 리팩터링 및 시큐리티 테스트 지원 추가

- AddressControllerTest를 WebMvcTest로 전환하여 실제 HTTP 요청/응답 기반 테스트 구현
- spring-security-test 의존성 추가로 인증/인가 테스트 지원
- 기존 Mockito 단위 테스트를 MockMvc 통합 테스트로 변경
- 네이버 OAuth2 사용자 인증 시나리오 및 주소지 생성 성공 케이스 검증

Refs: #45, #47

mochafreddo committed:
test: 주소지 생성 시 사용자 미존재 404 반환 테스트 추가

- 인증 정보에 해당하는 사용자를 찾을 수 없을 때 404 응답 및 에러 메시지 검증 테스트 추가
- DroniNotFoundException 처리 및 에러 응답 구조 검증

Refs: #45, #47

mochafreddo committed:
feat(address): 주소지 목록 조회 API 테스트 케이스 추가

- GET /api/v1/addresses 성공/빈 리스트/유저 없음(404) 시나리오 단위 테스트 추가
- UserAddress 엔티티의 createdAt/updatedAt 필드 설정 포함
- droniUserRepository, addressService 모킹 및 검증 로직 반영

Refs: #45, #47

mochafreddo committed:
test: AddressControllerTest 중복 코드 유틸 메서드화 및 테스트 가독성 개선

- 주소지 생성/조회 테스트에서 AddressSaveRequest, 인증 토큰, 감사 필드 설정 로직을 private 메서드로 분리하여 중복 제거
- 각 테스트에서 유틸 메서드 활용으로 코드 간결화 및 유지보수성 향상
- 기존 테스트 동작 및 검증 로직은 변경 없음

Refs: #45, #47

mochafreddo committed:
feat(address): UserAddress 생성 방식 create 팩토리 메서드로 통일 및 테스트/서비스 코드 리팩토링

UserAddress 엔티티에 create 정적 팩토리 메서드를 추가하고, 기존 builder 직접 호출 부분을 모두 create 메서드 사용으로 변경했습니다.
서비스 및 테스트 코드에서 UserAddress 생성 시 create 메서드를 활용하도록 리팩토링하여 생성 방식 일관성을 확보했습니다.

Refs: #45, #47

mochafreddo committed:
feat(address): 주소 기본(primary) 지정 로직 단순화 및 중첩 if 제거

- 기존 주소가 없을 경우, 새 주소를 무조건 기본(primary)으로 지정
- 새 주소가 기본(primary)로 지정될 경우, 기존 기본 주소를 해제
- 기존 주소에 기본(primary)이 없을 경우, 새 주소를 기본으로 지정
- 중첩된 if/else 구조를 단순화하여 가독성 및 유지보수성 개선

Refs: #45, #47

mochafreddo committed:
docs(address): AddressService 주요 메서드에 한글 Javadoc 추가

AddressService의 saveAddress, getUserAddresses 메서드에 한글 Javadoc을 추가하여 코드 문서화 및 가독성을 개선했습니다.

mochafreddo committed:
feat: `ResponseEntity` 제거 및 `ResponseStatus` 적용

- `ResponseEntity`를 사용하지 않고도 응답 상태를 명시적으로 설정할 수 있도록 `@ResponseStatus` 어노테이션을 적용했습니다.
- `@ResponseStatus(HttpStatus.CREATED)`를 통해 주소지 생성 시 201 Created를 반환하도록 했습니다.
- 기존 `@GetMapping`의 경우 기본적으로 200 OK를 반환하므로 별도의 어노테이션을 추가하지 않았습니다.
- `ResponseEntity`를 제거함에 따라 코드 가독성이 향상되었습니다.

Refs: #45, #47

mochafreddo committed:
feat(address): OAuth2UserPrincipal 기반 주소지 API 리팩터링 및 테스트 개선

- AddressController, AddressService에서 기존 userId 기반 로직을 OAuth2UserPrincipal 기반으로 변경
- DroniUserRepository 의존성 제거 및 서비스 내부에서 principal로 사용자 조회 일원화
- AddressControllerTest, AddressServiceTest에서 userId 대신 OAuth2UserPrincipal을 활용하도록 테스트 코드 전체 리팩터링
- 관련 불필요한 코드 및 검증 로직 정리
- 사용자 인증 정보 기반 예외 처리 및 반환값 검증 일관성 강화

Refs: #45, #47

mochafreddo committed:
test: AddressServiceTest 단위 테스트 명확화 및 불필요한 케이스 제거

- 주소 저장/조회 관련 단위 테스트의 DisplayName을 실제 동작 중심으로 명확하게 변경
- 불필요한 성공/실패 코드명 주석 제거 및 한글 설명으로 통일
- 주소 목록 조회 성공/빈 리스트 반환 테스트 케이스 제거 (비즈니스 로직 단순 조회는 테스트 생략 관례 반영)
- 테스트 코드 가독성 및 도메인 관례 개선

Refs: #45, #47

mochafreddo committed:
feat(address): AddressSaveRequest에 Bean Validation 적용 및 컨트롤러에 @Valid 추가

- AddressSaveRequest DTO에 @NotBlank, @Size, @Pattern 등 Bean Validation 어노테이션을 추가하여 입력값 검증을 강화함
- 국제전화 및 하이픈 없는 연락처 저장을 위해 E.164 표준 정규식 적용
- AddressController의 주소 생성 API에 @Valid 적용하여 요청값 자동 검증 처리

Refs: #45, #47

TaeTaeKim commented on src/main/java/droni/backend/api/address/entity/UserAddress.java:
```diff
@@ -0,0 +1,110 @@
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
+
+    public UserAddress(
```
요 부분 Entity의 AllArgConstructor는 사실 public 으로 두지 않기도 하고 사용처가 없으니 제거하는 건 어떨까요?

mochafreddo replied:
반영했습니다. (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/1f99e76dcee9d54fddbee00aa289d8f319b6ff30)

TaeTaeKim commented on src/main/java/droni/backend/api/address/dto/AddressSaveRequest.java:
```diff
@@ -0,0 +1,40 @@
+    @Size(max = 100, message = "상세주소는 100자 이하여야 합니다.")
+    private String address2;
+
+    public void setPrimary(boolean primary) {
```
이 Setter의 메소드는 어떤 이유때문에 있는 걸까요? Class Level 에 @Setter 가 들어가 있어서 질문드립니다!

mochafreddo replied:
`setPrimary`는 "기본 주소지" 변경할 때 사용하려고 만들었는데 다른 방법이 좋을까요?

mochafreddo replied:
불필요한 것이 맞아 제거했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/ed384ef90803b7ad2d880750af78b981f21f914f)

TaeTaeKim commented on src/main/java/droni/backend/api/address/dto/AddressSaveRequest.java:
```diff
@@ -0,0 +1,40 @@
+    @Size(max = 50, message = "주소 이름은 50자 이하여야 합니다.")
+    private String addressName;
+
+    private boolean isPrimary;
```
그리고 Java Jackson 의 Serialize 에서 해당 부분은 이슈가 있으니 확인해보시는게 좋을거에요!
https://00h0.tistory.com/93

mochafreddo replied:
수정했습니다. (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/ca95b76b7e9acd4181e0cc445dc79dc040b0801a)
확인 부탁드려요 :)

TaeTaeKim commented on src/main/java/droni/backend/api/address/service/AddressService.java:
```diff
@@ -0,0 +1,72 @@
+        boolean isNewAddressPrimary = request.isPrimary();
+
+        if (existingAddresses.isEmpty()) {
+            isNewAddressPrimary = true;
```
request의 isPrimary 필드를 변경하는건 별로일까요?

mochafreddo replied:
인바운드 요청 DTO는 변경을 지양한다고 합니다. 불변(immutable) 객체로 취급하고, 도메인/서비스 레이어에서 최종 상태를 도출해야 한다고 하네요 🤔

mochafreddo replied:
인바운드 요청 DTO를 변경하지 않는 것은 매우 중요한 개발 원칙입니다. 몇 가지 이유를 설명해 드릴게요.

### 1. **데이터 무결성(Data Integrity) 보장**

가장 중요한 이유입니다. DTO는 클라이언트로부터 받은 **"요청 원본 데이터"**를 나타냅니다. 이 데이터를 서비스 계층이나 다른 로직에서 임의로 변경하면, 요청의 본래 의도와 다른 데이터로 처리될 수 있습니다.

예를 들어, 사용자 회원가입 요청 DTO에 `password` 필드가 있다고 가정해 봅시다. 만약 서비스 로직에서 `password`를 해싱하기 전에 DTO 자체를 변경해서 `hashPassword`라는 새로운 필드를 추가한다면, 다른 코드에서 `password`와 `hashPassword`를 혼동하거나 잘못 참조할 가능성이 생깁니다. 원본 DTO는 최대한 순수한 상태로 유지하고, 변경이 필요한 경우 **새로운 객체**를 만들어 사용하거나 로직 내부의 **지역 변수**로 처리하는 것이 안전합니다.

---

### 2. **예측 가능성(Predictability)과 유지보수 용이성**

코드는 읽기 쉬워야 합니다. 만약 DTO가 여러 메서드를 거치면서 내용이 계속 바뀐다면, 개발자는 해당 DTO의 최종 상태를 예측하기 어려워집니다.

DTO를 불변(immutable) 객체로 다루면, 개발자는 DTO를 받았을 때 그 내용이 변하지 않는다는 것을 확신할 수 있습니다. 이는 코드의 흐름을 쉽게 추적하게 하고, 디버깅을 훨씬 수월하게 만듭니다. 복잡한 로직에서 예기치 않은 버그가 발생하는 것을 방지하는 데 큰 도움이 됩니다.

---

### 3. **경쟁 상태(Race Condition) 방지**

멀티스레드 환경에서는 DTO를 변경하는 것이 심각한 문제를 일으킬 수 있습니다. 여러 스레드가 하나의 DTO 인스턴스에 동시에 접근하여 값을 변경하려고 하면, 데이터가 꼬이거나 예상치 못한 결과가 발생할 수 있습니다.

특히 Spring Boot와 같은 환경에서는 서비스가 기본적으로 **싱글톤(Singleton)**으로 동작하기 때문에, 인스턴스 변수를 사용할 때 특히 주의해야 합니다. DTO는 메서드 호출마다 새로 생성되는 지역 객체이지만, 만약 DTO를 변경하는 로직을 다른 곳에서도 공유한다면 경쟁 상태의 위험이 커집니다. DTO를 불변으로 취급하면 이런 종류의 동시성 문제를 근본적으로 차단할 수 있습니다.

### 요약

요약하자면, 인바운드 DTO는 **"요청에 대한 스냅샷"**이라고 생각하는 게 좋습니다. 이 스냅샷은 절대 조작하면 안 됩니다. 만약 데이터에 어떤 처리를 가해야 한다면, DTO의 데이터를 기반으로 새로운 객체나 변수를 생성해서 사용하세요.

TaeTaeKim commented on src/test/java/droni/backend/api/address/controller/AddressControllerTest.java:
```diff
@@ -0,0 +1,234 @@
+    @MockBean
+    private AddressService addressService;
+
+    @MockBean
```
해당 부분은 뺴도 좋을 거 같습니다! 사용하지 않는 것 같아서요

mochafreddo replied:
JPA Auditing 때문에 `NoSuchBeanDefinitionException`등 오류가 발생해서 추가했습니다.
JPA가 제외된 상태에서 MVC 슬라이스가 실행될 수 있도록 stub을 제공받으려 사용했는데 더 좋은, 다른 방법이 있을까요? 🥹

TaeTaeKim replied:
음 제가 이해한 바로는 현재 Controller Test 에 있는 내용들이 사실은 Service Logic 검증에 들어가야하는 부분들이라 concept이 맞지 않아 말씀하신 내용들이 문제가 되지 않나 싶습니다.

"주소지 생성 성공", "인증 정보에 해당하는 사용자를 찾을 수 없을 때 404 반환" 이런 부분들이 결국에는 비즈니스 로직에 대한 검증으로 생각이 들어요. MockMVC테스트에서는 주로 consume, produce등에 대한 검증, Request Validation 동작에 대한 검증(전화번호 내용이 잘 검증되는지) 등등에 대한 검증이 주로 구현했던 것 같습니다.

현재 작성하신 테스트 내용을 보면  MockMVC가 아닌 SpringBOotTest로 통합테스트로 Presentation layer 부터 Persist Layer까지 테스트를 하는건 어떨까요?

TaeTaeKim commented on src/test/java/droni/backend/api/address/controller/AddressControllerTest.java:
```diff
@@ -0,0 +1,234 @@
+        return new TestingAuthenticationToken(testPrincipal, null, "ROLE_USER");
+    }
+
+    private void setAuditFields(UserAddress address, LocalDateTime time) {
```
해당 부분은 보니까 생성 팩토리에 created, update를 설정할 수 없어서 reflection method 로 만드신 것 같은데 mockito의 `spy` 를 찾아보시면 아마 쉽게 자바객체의 해동을 제어할 수 있을거에요!

mochafreddo replied:
반영했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/7403178001c7e79899294a5f79bea542b190200d)

TaeTaeKim reviewed (approved):
너무 리뷰가 많은 것 같아서... 딱히 중요한 변경들은 아니니 확인해보시고 resolve 하시면 될 거 같습니다.

TaeTaeKim commented on src/main/java/droni/backend/api/address/service/AddressService.java:
```diff
@@ -0,0 +1,72 @@
+    public UserAddress saveAddress(OAuth2UserPrincipal userPrincipal, AddressSaveRequest request) {
+        DroniUser user = findUserByPrincipal(userPrincipal);
+
+        List<UserAddress> existingAddresses = userAddressRepository.findByUser(user);
```
해당 레포지토리는 유저와 isPrimary를 같이 받아서 유저의 기본주소를 Optional 로 받는 걸로 수정하는 건 어떠세요?

mochafreddo replied:
반영했습니다! (https://github.com/DronicaPlatformService/droni-backend/pull/47/commits/40715d009148ea4286aeaea75cf99d47f2908bad)

mochafreddo replied:
```sql
CREATE UNIQUE INDEX ux_user_primary_true ON user_address(user_id) WHERE primary = true;
```
유니크 인덱스를 고려해 보라고 하더라구요. 참고해 주세요 :)

mochafreddo committed:
feat(address): UserAddress 엔티티에 주소 정보 수정 메서드 추가

UserAddress 엔티티에 주소명, 수령인, 연락처, 주소1/2를 수정할 수 있는 updateAddress 메서드를 추가했습니다. 기존 주소 정보 변경 시 활용할 수 있습니다.

Refs: #45, #47

mochafreddo committed:
refactor(entity): UserAddress 엔티티의 public AllArgsConstructor 제거

- 외부에서 직접 엔티티 생성 방지를 위해 모든 필드를 받는 public 생성자 삭제
- 객체 생성은 private 빌더 및 정적 팩토리 메서드(create)만 허용
- DDD 및 프로젝트 생성 패턴 가이드 준수

Refs: #45, #47

mochafreddo committed:
feat(address): AddressSaveRequest의 setPrimary 메서드 제거

AddressSaveRequest DTO에서 불필요한 setPrimary(boolean) 메서드를 삭제하여 코드의 명확성과 일관성을 개선했습니다.

Refs: #47

mochafreddo committed:
feat: 주소 기본값 필드명 및 타입 일관성 개선

- AddressSaveRequest/AddressResponse DTO의 isPrimary 필드를 Boolean 타입으로 변경하고, 요청 DTO에 @NotNull 추가
- UserAddress 엔티티의 기본 주소 여부 필드명을 isPrimary → primary로 변경하여 도메인/엔티티 네이밍 관례에 맞춤
- AddressService, 테스트 코드 등 관련 필드 접근자 및 생성자 파라미터를 모두 일관성 있게 수정
- Swagger 문서화용 @Schema(required) 어노테이션 추가로 API 명세 명확화

Refs: #47

mochafreddo committed:
feat(address): 기본 주소 설정 로직 리팩토링 및 가독성 개선

- 기본 주소 지정 여부 판단 로직을 `resolvePrimaryFlag` 메서드로 분리하여 중복 제거 및 가독성 향상
- 기존 주소가 없거나, 모든 주소가 비활성화된 경우 자동으로 기본 주소로 설정되도록 개선
- 기존 기본 주소가 있을 경우 새 기본 주소 저장 시 기존 기본 주소를 비활성화 처리

Refs: #47

mochafreddo committed:
feat(test): UserAddress 엔티티의 createdAt/updatedAt 설정 방식 Mockito doReturn으로 변경

UserAddress 테스트 객체의 생성/수정일 필드 설정에 기존 reflection을 사용하던 부분을 제거하고,
Mockito의 doReturn을 활용해 getter 메서드 반환값을 직접 제어하도록 개선하였습니다.
테스트 코드의 유지보수성과 가독성을 높이고, 엔티티 내부 구현 변경에도 더 안전하게 대응할 수 있습니다.

Refs: #47

mochafreddo committed:
feat(address): 기본 주소 판별 로직 리팩토링 및 테스트 개선

- UserAddressRepository에 existsByUser, findByUserAndPrimaryTrue 메서드 추가
- AddressService에서 기존 주소 조회 및 기본 주소 판별 로직을 existsByUser, findByUserAndPrimaryTrue 기반으로 변경
- 불필요한 resolvePrimaryFlag 메서드 제거
- AddressServiceTest에서 Repository 목킹 방식 변경 및 불필요한 객체 제거
- AddressControllerTest에서 doReturn 스텁 및 관련 import 제거

Refs: #47
